src/librbd/ImageWatcher.cc

```diff
@@ -982,6 +982,8 @@ void ImageWatcher<I>::handle_rewatch_complete(int r) {
     RWLock::RLocker owner_locker(m_image_ctx.owner_lock);
     if (m_image_ctx.exclusive_lock != nullptr) {
       // update the lock cookie with the new watch handle
+      // cookie exclusive_lock hold must be invalid after rewatch 
+      m_image_ctx.exclusive_lock->set_cookie_valid(false);
```

Comment:
Nit: can this invalidation of the cookie just be hidden within ```ManagedLock::reacquire_lock```?

Reply:
we can do this.
but from what now i can think of, it needs to define another parameters to reacquire_lock, just like following:
before:
``void reacquire_lock(Context *on_reacquired = nullptr);``
after:
``void reacquire_lock(Context *on_reacquired = nullptr, ***cookie_valid = true***);``

then explicity call 
`m_image_ctx.exclusive_lock->reacquire_lock(cookie_valid = false);`

is that wath you want?

Reply:
... I think this comment is pre-empted by the comment below.